core/src/main/scala/org/bitcoins/core/wallet/utxo/UTXOSpendingInfo.scala

```diff
@@ -17,26 +15,71 @@ sealed abstract class UTXOSpendingInfo {
+                                     scriptPubKey: ScriptPubKey,
+                                     signers: Seq[Sign],
+                                     hashType: HashType) extends BitcoinUTXOSpendingInfo
+
+
+/** This is the case where we are spending a [[org.bitcoins.core.protocol.script.WitnessScriptPubKeyV0 witness v0 script]]  */
+case class SegwitV0NativeUTXOSpendingInfo(outPoint: TransactionOutPoint,
+                                          amount: CurrencyUnit,
+                                          scriptPubKey: WitnessScriptPubKeyV0,
+                                          signers: Seq[Sign],
+                                          hashType: HashType,
+                                          scriptWitness: ScriptWitnessV0) extends BitcoinUTXOSpendingInfo
+
+/** This is the case were we are attempting to spend a [[org.bitcoins.core.protocol.script.P2SHScriptPubKey p2sh spk]] */
+case class P2SHSpendingInfo(outPoint: TransactionOutPoint,
+                            amount: CurrencyUnit,
+                            scriptPubKey: P2SHScriptPubKey,
+                            signers: Seq[Sign],
+                            hashType: HashType,
+                            redeemScript: ScriptPubKey) extends BitcoinUTXOSpendingInfo {
+  require(P2SHScriptPubKey(redeemScript) == output.scriptPubKey,
+    s"Given redeem script did not match hash in output script, " +
+      s"got=${P2SHScriptPubKey(redeemScript).scriptHash.hex}, " +
+      s"expected=${scriptPubKey.scriptHash.hex}")
 }
 
-case class BitcoinUTXOSpendingInfo(
-    outPoint: TransactionOutPoint,
-    output: TransactionOutput,
-    signers: Seq[Sign],
-    redeemScriptOpt: Option[ScriptPubKey],
-    scriptWitnessOpt: Option[ScriptWitness],
-    hashType: HashType)
-    extends UTXOSpendingInfo
+
+/** This is for the case we are spending a p2sh(p2w{pkh,sh}) script. This means that
+  * we have nested a [[org.bitcoins.core.protocol.script.WitnessScriptPubKeyV0 witness spk]]
+  * inside of a [[org.bitcoins.core.protocol.script.P2SHScriptPubKey p2sh spk]] */
+case class P2SHNestedSegwitV0UTXOSpendingInfo(outPoint: TransactionOutPoint,
+                                              amount: CurrencyUnit,
+                                                scriptPubKey: P2SHScriptPubKey,
+                                               signers: Seq[Sign],
+                                               hashType: HashType,
+                                               redeemScript: WitnessScriptPubKeyV0,
+                                               scriptWitness: ScriptWitnessV0
+) extends BitcoinUTXOSpendingInfo {
+  require(P2SHScriptPubKey(redeemScript) == output.scriptPubKey,
+    s"Given redeem script did not match hash in output script, " +
+      s"got=${P2SHScriptPubKey(redeemScript).scriptHash.hex}, " +
+      s"expected=${scriptPubKey.scriptHash.hex}")
+
+  //can we put an invariant here with scriptWitness and the witness spk redeemScript?
```

Comment:
Yes, I believe if you match on the `redeemScript` you should be able to add an invariant depending on whether it is pubkey or script